epuck_ros2_cpp/src/driver.cpp

```diff
@@ -0,0 +1,447 @@
+      geometry_msgs::msg::TransformStamped infrared_transform;
+      infrared_transform.header.stamp = this->now();
+      infrared_transform.header.frame_id = "base_link";
+      infrared_transform.child_frame_id = "ps" + std::to_string(i);
+      infrared_transform.transform.rotation = EPuckPublisher::euler_to_quaternion(0, 0, DISTANCE_SENSOR_ANGLE[i]);
+      infrared_transform.transform.translation.x = 0;
+      infrared_transform.transform.translation.y = 0;
+      infrared_transform.transform.translation.z = 0;
+      infrared_broadcasters[i]->sendTransform(infrared_transform);
+    }
+
+    // Static tf broadcaster: Range (ToF)
+    infrared_broadcasters[8] = std::make_unique<tf2_ros::StaticTransformBroadcaster>(this);
+    geometry_msgs::msg::TransformStamped infrared_transform;
+    infrared_transform.header.stamp = this->now();
+    infrared_transform.header.frame_id = "base_link";
+    infrared_transform.child_frame_id = "tof";
+    infrared_transform.transform.rotation = EPuckPublisher::euler_to_quaternion(0, 0, 0);
+    infrared_transform.transform.translation.x = 0;
+    infrared_transform.transform.translation.y = 0;
+    infrared_transform.transform.translation.z = 0;
+    infrared_broadcasters[8]->sendTransform(infrared_transform);
+
+    RCLCPP_INFO(this->get_logger(), "EPuck Driver has been initialized");
+    RCLCPP_INFO(this->get_logger(), "Driver mode: %s", type.c_str());
+  }
+
+  ~EPuckPublisher() { close(fh); }
+
+private:
+  void reset_odometry() {
+    prev_left_wheel_raw = 0;
+    prev_right_wheel_raw = 0;
+    odom_left_overflow = 0;
+    odom_right_overflow = 0;
+    prev_position_x = 0;
+    prev_position_y = 0;
+    prev_angle = 0;
+  }
+
+  rcl_interfaces::msg::SetParametersResult param_change_callback(std::vector<rclcpp::Parameter> parameters) {
+    auto result = rcl_interfaces::msg::SetParametersResult();
+    result.successful = true;
+
+    for (auto parameter : parameters) {
+      if (parameter.get_name() == "wheel_distance") {
+        reset_odometry();
+        wheel_distance = static_cast<float>(parameter.as_double());
+      } else if (parameter.get_name() == "wheel_radius") {
+        reset_odometry();
+        wheel_radius = static_cast<float>(parameter.as_double());
+      }
+
+      RCLCPP_INFO(this->get_logger(), "Parameter '%s' has changed to %s", parameter.get_name().c_str(),
+                  parameter.value_to_string().c_str());
+    }
+
+    return result;
+  }
+
+  static float intensity_to_distance(int p_x) {
+    for (unsigned int i = 0; i < INFRARED_TABLE.size() - 1; i++)
+      if (INFRARED_TABLE[i][1] >= p_x && INFRARED_TABLE[i + 1][1] < p_x) {
+        const float b_x = INFRARED_TABLE[i][1];
+        const float b_y = INFRARED_TABLE[i][0];
+        const float a_x = INFRARED_TABLE[i + 1][1];
+        const float a_y = INFRARED_TABLE[i + 1][0];
+        const float p_y = ((b_y - a_y) / (b_x - a_x)) * (p_x - a_x) + a_y;
+        return p_y;
+      }
```

Comment:
Here we usually add brackets: https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style#cs360-single-if-while-or-for-statement-must-be-written-without-brackets-the-statement-part-should-be-put-on-a-separate-line-in-case-of-nested-block-keep-brackets